include/gz/math/TimeVaryingVolumetricGrid.hh

```diff
@@ -0,0 +1,152 @@
+#include <gz/math/Vector3.hh>
+
+#include <map>
+#include <utility>
+#include <vector>
+
+namespace gz
+{
+namespace math
+{
+
+/// \brief A grid with interpolation where time can be varied. This class has
+/// no implementation to allow for different strategies.
+/// \sa InMemoryTimeVaryingVolumetricGrid for an example
+/// specialization that loads all data into memory. To use this class you should
+/// use `CreateSession` so you can step through time.
+template<typename T, typename V, typename S, typename P>
+class TimeVaryingVolumetricGrid
+{
+  /// \brief Creates a session. Call this before querying the interface.
+  public: S CreateSession() const;
+
+  /// \brief Steps the session to a fixed time step
+  /// \param _session - The session to be stepped
+  /// \param _time - The time at which the session should be set.
+  /// Returns the session cursor if the step was successful. If there was
+  /// no more data return nullopt.
+  public: std::optional<S> StepTo(const S &_session, const T &_time) const;
+
+  /// \brief Looks up a given point in time.
+  /// \param _session - The session with the time stamp to be looked up.
+  /// \param _pos - The position of the point we want to query
+  public: std::optional<V> LookUp(const S &_session, const Vector3<P> &_pos)
+    const;
+};
+
+/// \brief Specialization of TimeVaryingVolumetricGrid which loads the whole of
+/// the dataset into memory. To construct this class use
+/// `InMemoryTimeVaryingVolumetricGridFactory`
+template<typename T, typename V, typename P>
+class TimeVaryingVolumetricGrid<T, V, InMemorySession<T, P>, P>
+{
+  /// \brief Documentation Inherited
+  public: InMemorySession<T, V> CreateSession() const
+  {
+    return indices.CreateSession();
+  }
+
+  /// \brief Documentation Inherited
+  public: std::optional<InMemorySession<T, P>>
+    StepTo(const InMemorySession<T, double> &_session, const T &_time) const
+  {
+    return indices.StepTo(_session, _time);
+  }
+
+  /// \brief Looks up a given point. If the point lies in between two time
+  /// frames then it performs spatio-temporal linear interpolation.
+  /// Returns nullopt if the data is out of range.
+  public: std::optional<V>
+    LookUp(const InMemorySession<T, P> &_session, const Vector3<P> &_pos)
+    const
+  {
+    auto points = indices.LookUp(_session, _pos);
+    std::optional<V> result = indices.EstimateQuadrilinear(
+      _session,
+      points,
+      values,
+      values,
+      _pos);
+    return result;
+  }
+
+  /// Buffer for values being stored
+  private: std::vector<V> values;
+
+  /// Index table for fast lookup
+  private: TimeVaryingVolumetricGridLookupField
+    <T, V, InMemorySession<T, P>> indices;
+
+  template<typename U, typename S, typename X>
+  friend class InMemoryTimeVaryingVolumetricGridFactory;
+};
+
+/// \brief Alias for Specialization of TimeVaryingVolumetricGrid which loads
+/// the whole of the dataset into memory.
+template<typename T, typename V, typename P>
```

Comment:
@arjo129 nit^2: it would be nice and convenient to use T as default for V and P. Then we can go from `InMemoryTimeVaryingVolumetricGrid<double, double, double>` to `InMemoryTimeVaryingVolumetricGrid<double>`. Same elsewhere applicable.